app/services/agreements/base_seeder.rb

```diff
@@ -0,0 +1,59 @@
+module Agreements
+  class BaseSeeder
+    def initialize(rails_env: Rails.env, yaml_path: 'config')
+      @rails_env = rails_env
+      @yaml_path = yaml_path
+    end
+
+    def run
+      records.each do |key, config|
+        config = process_config(key, config)
+        record = record_class.find_or_initialize_by(
+          primary_attribute_bundle(config)
+        )
+        record.assign_attributes(config)
+        record.save!
+      end
+      after_seed
+    end
+
+    private
+
+    attr_reader :rails_env, :yaml_path
+
+    # The following methods need to be defined in the child class
```

Comment:
I wanted to document these somewhere but since Ruby will naturally raise a `NoMethodError` if the methods are undefined I didn't think most of them needed to be actually defined to raise the error, hence the extended comment.

I might tweak the examples to show something more realistic rather than the `raise` I originally added.